src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationAndGeneratorDriverTranslationAction.cs

```diff
@@ -5,43 +5,39 @@
 using System.Threading;
 using System.Threading.Tasks;
 
-namespace Microsoft.CodeAnalysis
+namespace Microsoft.CodeAnalysis;
+
+internal partial class SolutionCompilationState
 {
-    internal partial class SolutionCompilationState
+    /// <summary>
+    /// Represents a change that needs to be made to a <see cref="Compilation"/>, <see cref="GeneratorDriver"/>, or both in response to
+    /// some user edit.
+    /// </summary>
+    private abstract partial class CompilationAndGeneratorDriverTranslationAction
     {
+        public abstract Task<Compilation> TransformCompilationAsync(Compilation oldCompilation, CancellationToken cancellationToken);
```

Comment:
i far preferred subclasses having to be intentional about what they should do here.  i already did something similar for CanUpdateCompilationWithStaleGeneratedTreesIfGeneratorsGiveSameOutput  previously.

Reply:
now, the only non-abstract method is TryMergeWithPrior.  And that one is fine to stay virtual as it's only used as a perf optimization, not a correctness one.